File: frontend/lib/core/models/product.dart
class Product {
  final String id;
  final String name;
  final String? description;
  final String? brand;
  final String? category;
  final String sku;
  final double price;
  final double? salePrice;
  final int stockQty;
  final double weightKg;
  final String? imageUrl;
  final List<String> images;
  final bool isActive;

  Product({
    required this.id,
    required this.name,
    this.description,
    this.brand,
    this.category,
    required this.sku,
    required this.price,
    this.salePrice,
    required this.stockQty,
    required this.weightKg,
    this.imageUrl,
    this.images = const [],
    this.isActive = true,
  });

  double get effectivePrice => salePrice ?? price;

  bool get isOnSale => salePrice != null && salePrice! < price;

  bool get inStock => stockQty > 0;

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      brand: json['brand'] as String?,
      category: json['category'] as String?,
      sku: json['sku'] as String,
      price: (json['price'] as num).toDouble(),
      salePrice: (json['sale_price'] as num?)?.toDouble(),
      stockQty: json['stock_qty'] as int? ?? 0,
      weightKg: (json['weight_kg'] as num?)?.toDouble() ?? 0.0,
      imageUrl: json['image_url'] as String?,
      images: (json['images'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      isActive: json['is_active'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'description': description,
        'brand': brand,
        'category': category,
        'sku': sku,
        'price': price,
        'sale_price': salePrice,
        'stock_qty': stockQty,
        'weight_kg': weightKg,
        'image_url': imageUrl,
        'images': images,
        'is_active': isActive,
      };
}

File: backend/app/schemas/product.py
"""Product Pydantic schemas."""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class ProductBase(BaseModel):
    sku: str
    name: str
    description: str | None = None
    brand: str | None = None
    category: str
    sub_category: str | None = None
    price: float = Field(gt=0)
    sale_price: float | None = None
    weight_grams: int | None = None
    images: list[str] | None = None
    specifications: dict | None = None


class ProductCreate(ProductBase):
    cost_price: float | None = None
    stock_quantity: int = 0


class ProductUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    brand: str | None = None
    category: str | None = None
    sub_category: str | None = None
    price: float | None = Field(None, gt=0)
    sale_price: float | None = None
    cost_price: float | None = None
    stock_quantity: int | None = None
    weight_grams: int | None = None
    images: list[str] | None = None
    specifications: dict | None = None
    is_active: bool | None = None


class ProductResponse(ProductBase):
    id: uuid.UUID
    cost_price: float | None = None
    stock_quantity: int
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class ProductCompatibilityCreate(BaseModel):
    product_id: uuid.UUID
    vessel_make: str | None = None
    vessel_model: str | None = None
    year_from: int | None = None
    year_to: int | None = None
    engine_make: str | None = None
    engine_model: str | None = None
    notes: str | None = None


class ProductCompatibilityResponse(ProductCompatibilityCreate):
    id: uuid.UUID

    model_config = {"from_attributes": True}


class CompatibilityCheck(BaseModel):
    is_compatible: bool
    message: str
